File: src/domain/services/kinhDich/nguHanhClassifier.ts
// src/domain/services/kinhDich/nguHanhClassifier.ts
// Ngũ Hành (Five Elements) interaction classifier for lower/upper trigrams.
// Pure domain — no I/O imports.

// ── Types ──────────────────────────────────────────────────────────────────────

/** The five elemental phases of Ngũ Hành (Wu Xing). */
export type NguHanh = "Kim" | "Moc" | "Thuy" | "Hoa" | "Tho";

/** Dynamic relationship between lower and upper trigram elements. */
export type NguHanhDynamic =
  | "TUONG_SINH"  // Mutual generation (lower generates upper, or upper generates lower)
  | "TUONG_KHAC"  // Mutual destruction (one controls/overcomes the other)
  | "SAME"        // Same element — reinforcement
  | "NEUTRAL";    // No special interaction

export interface NguHanhResult {
  lower: NguHanh;
  upper: NguHanh;
  dynamic: NguHanhDynamic;
  /** Score contribution: +0.3 (tuong sinh lower→upper), +0.2 (tuong sinh upper→lower),
   *  -0.3 (tuong khac), +0.1 (same), 0.0 (neutral) */
  score: number;
  /** Vietnamese plain-text interpretation for this interaction. */
  interpretation: string;
}

// ── Elemental cycles ──────────────────────────────────────────────────────────

/**
 * Generation (Tương Sinh / 相生) cycle — "A gives birth to B":
 *   Mộc → Hỏa → Thổ → Kim → Thủy → Mộc
 */
const GENERATION: Record<NguHanh, NguHanh> = {
  Moc: "Hoa",
  Hoa: "Tho",
  Tho: "Kim",
  Kim: "Thuy",
  Thuy: "Moc",
};

/**
 * Destruction (Tương Khắc / 相克) cycle — "A controls B":
 *   Mộc → Thổ → Thủy → Hỏa → Kim → Mộc
 */
const DESTRUCTION: Record<NguHanh, NguHanh> = {
  Moc: "Tho",
  Tho: "Thuy",
  Thuy: "Hoa",
  Hoa: "Kim",
  Kim: "Moc",
};

// ── Public API ─────────────────────────────────────────────────────────────────

/**
 * Classify the Ngũ Hành interaction between lower and upper trigram elements.
 *
 * Priority order (checked in sequence):
 * 1. Same element       → SAME,       +0.1
 * 2. Lower generates upper (GENERATION[lower] === upper) → TUONG_SINH, +0.3
 * 3. Upper generates lower (GENERATION[upper] === lower) → TUONG_SINH, +0.2
 * 4. Lower destroys upper  (DESTRUCTION[lower] === upper) → TUONG_KHAC, -0.3
 * 5. Upper destroys lower  (DESTRUCTION[upper] === lower) → TUONG_KHAC, -0.3
 * 6. Otherwise             → NEUTRAL,  0.0
 *
 * @param lowerElement - Ngũ Hành element of the lower trigram
 * @param upperElement - Ngũ Hành element of the upper trigram
 */
export function classifyNguHanh(
  lowerElement: NguHanh,
  upperElement: NguHanh,
): NguHanhResult {
  let dynamic: NguHanhDynamic;
  let score: number;
  let interpretation: string;

  if (lowerElement === upperElement) {
    dynamic = "SAME";
    score = 0.1;
    interpretation = "Cung hanh — tang cuong xu huong hien tai";
  } else if (GENERATION[lowerElement] === upperElement) {
    dynamic = "TUONG_SINH";
    score = 0.3;
    interpretation = "Noi luc sinh ngoai luc — xu huong ben vung";
  } else if (GENERATION[upperElement] === lowerElement) {
    dynamic = "TUONG_SINH";
    score = 0.2;
    interpretation = "Ngoai luc ho tro noi luc — moi truong thuan loi";
  } else if (DESTRUCTION[lowerElement] === upperElement) {
    dynamic = "TUONG_KHAC";
    score = -0.3;
    interpretation = "Noi luc khac ngoai luc — mau thuan noi tai";
  } else if (DESTRUCTION[upperElement] === lowerElement) {
    dynamic = "TUONG_KHAC";
    score = -0.3;
    interpretation = "Ngoai luc triet tieu noi luc — ap luc ben ngoai";
  } else {
    dynamic = "NEUTRAL";
    score = 0.0;
    interpretation = "Trung tinh — khong tuong tac dac biet";
  }

  return { lower: lowerElement, upper: upperElement, dynamic, score, interpretation };
}

File: apps/kinh-dich-service/pkg/primitive/ngu_hanh_classifier/ngu_hanh_classifier.go
// Package ngu_hanh_classifier provides pure functions for classifying the
// Ngu Hanh (Five Elements) interaction between lower and upper trigram elements.
//
// Fence-A: pure compute only — lookup tables embedded inline, no I/O.
// Zero imports from application, interface, infrastructure, or module layers.
package ngu_hanh_classifier

// NguHanh represents one of the five elements.
type NguHanh string

const (
	Kim  NguHanh = "Kim"
	Moc  NguHanh = "Moc"
	Thuy NguHanh = "Thuy"
	Hoa  NguHanh = "Hoa"
	Tho  NguHanh = "Tho"
)

// NguHanhDynamic represents the dynamic relationship between two elements.
type NguHanhDynamic string

const (
	TUONG_SINH NguHanhDynamic = "TUONG_SINH" // Generation cycle
	TUONG_KHAC NguHanhDynamic = "TUONG_KHAC" // Destruction cycle
	SAME       NguHanhDynamic = "SAME"       // Same element
	NEUTRAL    NguHanhDynamic = "NEUTRAL"    // No special interaction or unknown
)

// NguHanhResult contains the classification result.
type NguHanhResult struct {
	Lower          NguHanh        `json:"lower"`
	Upper          NguHanh        `json:"upper"`
	Dynamic        NguHanhDynamic `json:"dynamic"`
	Score          float64        `json:"score"`
	Interpretation string         `json:"interpretation"`
}

// generation maps each element to the element it generates.
// Five-element generation cycle: Moc -> Hoa -> Tho -> Kim -> Thuy -> Moc
var generation = map[NguHanh]NguHanh{
	Moc:  Hoa,
	Hoa:  Tho,
	Tho:  Kim,
	Kim:  Thuy,
	Thuy: Moc,
}

// destruction maps each element to the element it destroys.
// Five-element destruction cycle: Moc -> Tho -> Thuy -> Hoa -> Kim -> Moc
var destruction = map[NguHanh]NguHanh{
	Moc:  Tho,
	Tho:  Thuy,
	Thuy: Hoa,
	Hoa:  Kim,
	Kim:  Moc,
}

// validNguHanh is a set of valid NguHanh values for input validation.
var validNguHanh = map[string]NguHanh{
	"Kim":  Kim,
	"Moc":  Moc,
	"Thuy": Thuy,
	"Hoa":  Hoa,
	"Tho":  Tho,
}

// ClassifyNguHanh classifies the Ngu Hanh interaction between lower and upper
// trigram elements.
//
// Returns NguHanhResult with dynamic relationship, score, and interpretation.
//
// Unknown/invalid element strings return NEUTRAL (score 0.0) — documented choice:
// returning a neutral classification is safer than throwing for downstream
// pipeline callers that may receive raw, unvalidated market data strings.
func ClassifyNguHanh(lower, upper string) NguHanhResult {
	// Validate inputs
	lowerEl, lowerOK := validNguHanh[lower]
	upperEl, upperOK := validNguHanh[upper]

	// Unknown input -> NEUTRAL (documented choice: no throw)
	if !lowerOK || !upperOK {
		// Default to Tho for display if unknown
		if !lowerOK {
			lowerEl = Tho
		}
		if !upperOK {
			upperEl = Tho
		}
		return NguHanhResult{
			Lower:          lowerEl,
			Upper:          upperEl,
			Dynamic:        NEUTRAL,
			Score:          0.0,
			Interpretation: "Phan tu khong xac dinh — trung tinh mac dinh",
		}
	}

	var dynamic NguHanhDynamic
	var score float64
	var interpretation string

	switch {
	case lowerEl == upperEl:
		dynamic = SAME
		score = 0.1
		interpretation = "Cung hanh — tang cuong xu huong hien tai"
	case generation[lowerEl] == upperEl:
		dynamic = TUONG_SINH
		score = 0.3
		interpretation = "Noi luc sinh ngoai luc — xu huong ben vung"
	case generation[upperEl] == lowerEl:
		dynamic = TUONG_SINH
		score = 0.2
		interpretation = "Ngoai luc ho tro noi luc — moi truong thuan loi"
	case destruction[lowerEl] == upperEl:
		dynamic = TUONG_KHAC
		score = -0.3
		interpretation = "Noi luc khac ngoai luc — mau thuan noi tai"
	case destruction[upperEl] == lowerEl:
		dynamic = TUONG_KHAC
		score = -0.3
		interpretation = "Ngoai luc triet tieu noi luc — ap luc ben ngoai"
	default:
		dynamic = NEUTRAL
		score = 0.0
		interpretation = "Trung tinh — khong tuong tac dac biet"
	}

	return NguHanhResult{
		Lower:          lowerEl,
		Upper:          upperEl,
		Dynamic:        dynamic,
		Score:          score,
		Interpretation: interpretation,
	}
}
